man/man6/stackandconquer.6
'\" t
.\" ** The above line should force tbl to be a preprocessor **
.\" Man page for StackAndConquer
.\"
.\" Copyright (C), 2018, Thorsten Roth
.\"
.\" You may distribute under the terms of the GNU General Public
.\" License as specified in the file COPYING that comes with the man
.\" distribution.
.\"
.\" Mon Jan  01 20:15:00 CEST 2018  ElThoro <elthoro@gmx.de>
.\"
.TH STACKANDCONQUER 6 "2019-08-31" "Thorsten Roth" "StackAndConquer Manual"
.SH NAME
StackAndConquer \- Tower conquest board game
.SH SYNOPSIS
\fBstackandconquer\fP [\fIOption\fP] or/and [\fIFile/Folder\fP]
.SH DESCRIPTION
StackAndConquer is a challenging tower conquest board game.
.SH Options
.TP
\fB\-\-debug\fP
Start debug mode.
.TP
\fB\-\-genboard\fP
Convert *.in file into *.stackboard game board file.
.TP
\fBFile/Folder\fP
Load save game (.stacksav) or if option \-\-genboard is used together with
input file/folder and optional output folder generate a board file.
.SH FILES
.TP
.I /usr/share/stackandconquer/cpu
CPU scripts
.TP
.I /usr/share/stackandconquer/boards
Game boards
.SH BUGS
Bug tracker:

https://codeberg.org/ElTh0r0/stackandconquer/issues


debian/changelog
stackandconquer (0.11.1-0) noble; urgency=low

  [ Thorsten Roth ]
  * Add Russian translation - thanks to aaly11
  * CI fixes

 -- Thorsten Roth <elthoro@gmx.de>  Sat, 07 Jul 2025 14:58:00 +0200

stackandconquer (0.11.0-0) jammy; urgency=low

  [ Thorsten Roth ]
  * Extend CPU interface
  * Add Minimax and MonteCarlo CPU - thanks to maksverver
  * Drop support for ReactOS (>= Qt 5.12 required now)
  * Update Dutch translation - thanks to maksverver
  * Integrate into Transifex for translations

 -- Thorsten Roth <elthoro@gmx.de>  Sun, 10 Nov 2024 12:49:00 +0200

stackandconquer (0.10.1-0) jammy; urgency=low

  [ Thorsten Roth ]
  * Add support for dark scheme on Windows with Qt 6.5
  * Add Japanese and Polish rules

 -- Thorsten Roth <elthoro@gmx.de>  Mon, 20 May 2024 14:57:00 +0200

stackandconquer (0.10.0-0) focal; urgency=low

  [ Thorsten Roth ]
  * Add board settings dialog
  * Updated Italian translation - thanks to davi92
  * Add txt to json board file converter
  * Fix for loading user CPU scripts
  * ESLint, Prettier JS code

 -- Thorsten Roth <elthoro@gmx.de>  Sun, 01 May 2022 14:29:00 +0200

stackandconquer (0.9.0-0) focal; urgency=low

  [ Thorsten Roth ]
  * Improve CPU opponent
  * Dynamic scale change
  * Icon for dark background
  * Check for legal moves
  * Implement direction change
  * Save game fixes
  * Configurable stone color
  * Rewrite of board handling
  * Update Dutch translation - thanks to Elbert
  * Packaging changes
  * Support for Qt 6

 -- Thorsten Roth <elthoro@gmx.de>  Sat, 27 Jun 2021 15:27:00 +0200

stackandconquer (0.8.2-0) bionic; urgency=low

  [ Thorsten Roth ]
  * Added Italian translation - thanks to davi92
  * Some internal code improvements.

 -- Thorsten Roth <elthoro@gmx.de>  Fri, 27 Mar 2020 15:29:00 +0200

stackandconquer (0.8.1-0) xenial; urgency=low

  [ Thorsten Roth ]
  * Added Dutch translation - thanks to Vistaus
  * Improved CPU opponent
  * Added macOS build (untested)
  * Added AppImage

 -- Thorsten Roth <elthoro@gmx.de>  Sun, 11 Mar 2018 15:40:00 +0200

stackandconquer (0.8.0-2) xenial; urgency=low

  [ Thorsten Roth ]
  * Added rules.
  * Load/save game.
  * Fixed cmd argument handling.

 -- Thorsten Roth <elthoro@gmx.de>  Wed, 08 Nov 2017 16:23:00 +0200

stackandconquer (0.7.0-0) xenial; urgency=low

  [ Thorsten Roth ]
  * Improved CPU
  * Bug fixes

 -- Thorsten Roth <elthoro@gmx.de>  Tue, 19 Sep 2017 20:11:00 +0200

stackandconquer (0.6.0-0) xenial; urgency=low

  [ Thorsten Roth ]
  * Added JS engine for CPU scripts and removed plugin interface.
  * Minor gui changes and graphic fixes.
  * Added fallback menu icons.
  * Change gui language without restarting the game.
  * Moved project from Launchpad to GitHub.

 -- Thorsten Roth <elthoro@gmx.de>  Mon, 17 Apr 2017 20:25:00 +0200

stackandconquer (0.5.0-0) xenial; urgency=low

  [ Thorsten Roth ]
  * Added CPU plugin interface.
  * Highlight moves.

 -- Thorsten Roth <elthoro@gmx.de>  Sun, 13 Nov 2016 17:26:00 +0200

stackandconquer (0.4.0-0) xenial; urgency=low

  [ Thorsten Roth ]
  * Check for possible moves.
  * Check if previou move is reverted.
  * Pretty debug output.

 -- Thorsten Roth <elthoro@gmx.de>  Sun, 29 May 2016 01:29:00 +0200

stackandconquer (0.3.0-0) wily; urgency=low

  [ Thorsten Roth ]
  * Settings dialog.
  * Added German translation.

 -- Thorsten Roth <elthoro@gmx.de>  Sat, 16 Apr 2016 19:40:00 +0200
 
stackandconquer (0.2.0-0) wily; urgency=low

  [ Thorsten Roth ]
  * Improved graphics.

 -- Thorsten Roth <elthoro@gmx.de>  Fri, 08 Apr 2016 13:30:00 +0200

stackandconquer (0.1.0-0) wily; urgency=low

  [ Thorsten Roth ]
  * Initial commit. Main game logic implemented.

 -- Thorsten Roth <elthoro@gmx.de>  Thu, 07 Apr 2016 13:29:00 +0200


stackandconquer.cpp
// SPDX-FileCopyrightText: 2015-2025 Thorsten Roth
// SPDX-License-Identifier: GPL-3.0-or-later

#include "./stackandconquer.h"

#include <QApplication>
#include <QDebug>
#include <QDesktopServices>
#include <QFileDialog>
#include <QGraphicsView>
#include <QGridLayout>
#include <QLabel>
#include <QLibraryInfo>
#include <QMessageBox>
#include <QResizeEvent>
#include <QSlider>
#include <QTabWidget>
#include <QTextEdit>
#if QT_VERSION >= QT_VERSION_CHECK(6, 5, 0)
#include <QStyleHints>
#endif

#include "./game.h"
#include "ui_stackandconquer.h"

StackAndConquer::StackAndConquer(const QDir &userDataPath,
                                 const QStringList &sListArgs, QWidget *pParent)
    : QMainWindow(pParent),
      m_pUi(new Ui::StackAndConquer),
      m_userDataDir(userDataPath),
      m_sCurrLang(QString()),
      m_pSettings(Settings::instance()),
      m_pGame(nullptr),
      // Size is based on default grid size of 70!
      m_DefaultSize(600, 480) {
  m_pUi->setupUi(this);

  QString sIconTheme;
#if QT_VERSION >= QT_VERSION_CHECK(6, 5, 0)
  qDebug() << "Detected color scheme:"
           << QGuiApplication::styleHints()->colorScheme();
  if (Qt::ColorScheme::Dark == QGuiApplication::styleHints()->colorScheme()) {
    sIconTheme = QStringLiteral("dark");
  } else if (Qt::ColorScheme::Light ==
             QGuiApplication::styleHints()->colorScheme()) {
    sIconTheme = QStringLiteral("light");
  }
#endif
  // If < Qt 6.5 or if Qt::ColorScheme::Unknown was returned
  if (sIconTheme.isEmpty()) {
    // If window is darker than text
    if (this->window()->palette().window().color().lightnessF() <
        this->window()->palette().windowText().color().lightnessF()) {
      sIconTheme = QStringLiteral("dark");
    } else {
      sIconTheme = QStringLiteral("light");
    }
  }
  QIcon::setThemeName(sIconTheme);

  m_pSettingsDialog = new SettingsDialog(this, m_userDataDir.absolutePath());
  connect(m_pSettingsDialog, &SettingsDialog::newGame, this,
          &StackAndConquer::startNewGame);
  connect(m_pSettingsDialog, &SettingsDialog::changeLang, this,
          &StackAndConquer::loadLanguage);
  connect(this, &StackAndConquer::updateUiLang, m_pSettingsDialog,
          &SettingsDialog::updateUiLang);

  this->loadLanguage(m_pSettings->getGuiLanguage());
  this->setupMenu();
  this->setupGraphView();
  this->checkCmdArgs(sListArgs);
}

StackAndConquer::~StackAndConquer() = default;

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::checkCmdArgs(const QStringList &sListArgs) {
  // Open savegame from command line
  QString sSavegame(QLatin1String(""));
  if (!sListArgs.isEmpty()) {
    // Load save game
    if (sListArgs.at(0).endsWith(Settings::SAVE_FILE_EXT,
                                 Qt::CaseInsensitive)) {
      if (QFile::exists(sListArgs.at(0))) {
        sSavegame = sListArgs.at(0);
      } else {
        qWarning() << "Specified savegame not found:" << sListArgs.at(0);
        QMessageBox::warning(
            this, tr("Warning"),
            tr("Specified file not found:") + "\n" + sListArgs.at(0));
      }
    }
  }
  this->startNewGame(sSavegame);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::setupMenu() {
  // Game menu
  m_pUi->action_NewGame->setShortcut(QKeySequence::New);
  connect(m_pUi->action_NewGame, &QAction::triggered, this,
          [this]() { this->startNewGame(QString()); });
  m_pUi->action_LoadGame->setShortcut(QKeySequence::Open);
  connect(m_pUi->action_LoadGame, &QAction::triggered, this,
          &StackAndConquer::loadGame);
  m_pUi->action_SaveGame->setShortcut(QKeySequence::Save);
  connect(m_pUi->action_SaveGame, &QAction::triggered, this,
          &StackAndConquer::saveGame);
  connect(m_pUi->action_Preferences, &QAction::triggered, m_pSettingsDialog,
          &SettingsDialog::show);
  m_pUi->action_Quit->setShortcut(QKeySequence::Quit);
  connect(m_pUi->action_Quit, &QAction::triggered, this,
          &StackAndConquer::close);

  // Help menu
  connect(m_pUi->action_Rules, &QAction::triggered, this,
          &StackAndConquer::showRules);
  connect(m_pUi->action_ReportBug, &QAction::triggered, this,
          &StackAndConquer::reportBug);
  connect(m_pUi->action_Info, &QAction::triggered, this,
          &StackAndConquer::showInfoBox);

  m_pZoomSlider = new QSlider(Qt::Orientation::Horizontal, this);
  m_pZoomSlider->setMinimum(m_pSettings->getDefaultGridSize());
  m_pZoomSlider->setMaximum(m_pSettings->getMaxGridSize());
  m_pZoomSlider->setSingleStep(5);
  m_pZoomSlider->setTickInterval(10);
  m_pZoomSlider->setTracking(false);
  m_pZoomSlider->setMaximumWidth(200);
  m_pZoomSlider->setValue(m_pSettings->getGridSize());
  m_pUi->statusBar->addPermanentWidget(m_pZoomSlider);
  connect(m_pZoomSlider, &QSlider::valueChanged, this,
          &StackAndConquer::zoomChanged);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::setupGraphView() {
  m_pGraphView = new QGraphicsView(this);
  // Set mouse tracking to true, otherwise mouse move event
  // for the *scene* is only triggered on a mouse click!
  // QGraphicsView forwards the event to the scene.
  m_pGraphView->setMouseTracking(true);

  // Transform coordinate system to "isometric" view
  QTransform transfISO;
  transfISO = transfISO.scale(1.0, 0.5).rotate(45);
  m_pGraphView->setTransform(transfISO);
  this->setCentralWidget(m_pGraphView);

  m_pFrame = new QFrame(m_pGraphView);
  m_pLayout = new QGridLayout;
  m_pLayout->setVerticalSpacing(0);
  QPixmap iconWin(QStringLiteral(":/img/trophy.png"));
  if (m_pSettings->getBgColor().lightnessF() < 0.5) {
    iconWin = QStringLiteral(":/img/trophy2.png");
  }

  for (int i = 0; i < m_pSettings->getMaxNumOfPlayers(); i++) {
    m_pLblsPlayerName << new QLabel(QStringLiteral("Player"));
    m_pLblsPlayerName.last()->setStyleSheet(QStringLiteral("color: ") +
                                            m_pSettings->getTextColor().name());
    QSizePolicy sp_retain = m_pLblsPlayerName.last()->sizePolicy();
    sp_retain.setRetainSizeWhenHidden(true);
    m_pLblsPlayerName.last()->setSizePolicy(sp_retain);

    m_pLblsStoneIcon << new QLabel();
    this->drawPlayerIcon(i);
    m_pLblsStoneIcon.last()->setSizePolicy(sp_retain);
    m_pLblsStonesLeft << new QLabel(QStringLiteral("99"));
    m_pLblsStonesLeft.last()->setStyleSheet(QStringLiteral("color: ") +
                                            m_pSettings->getTextColor().name());
    m_pLblsStonesLeft.last()->setSizePolicy(sp_retain);

    m_pLblsWinIcon << new QLabel();
    m_pLblsWinIcon.last()->setPixmap(iconWin);
    m_pLblsWinIcon.last()->setAlignment(Qt::AlignCenter);
    m_pLblsWinIcon.last()->setSizePolicy(sp_retain);
    m_pLblsWon << new QLabel(QStringLiteral("0"));
    m_pLblsWon.last()->setStyleSheet(QStringLiteral("color: ") +
                                     m_pSettings->getTextColor().name());
    m_pLblsWon.last()->setSizePolicy(sp_retain);

    if (0 == i % 2) {
      // addWidget(*widget, row, column, rowspan, colspan)
      m_pLayout->addWidget(m_pLblsPlayerName.last(), i + ((i / 2) * 4), 0, 1,
                           2);
      m_pLayout->addWidget(m_pLblsStoneIcon.last(), i + 1 + ((i / 2) * 4), 0, 1,
                           1);
      m_pLayout->addWidget(m_pLblsStonesLeft.last(), i + 1 + ((i / 2) * 4), 1,
                           1, 1);
      m_pLayout->addWidget(m_pLblsWinIcon.last(), i + 2 + ((i / 2) * 4), 0, 1,
                           1);
      m_pLayout->addWidget(m_pLblsWon.last(), i + 2 + ((i / 2) * 4), 1, 1, 1);
    } else {
      m_pLayout->addWidget(m_pLblsPlayerName.last(), i - 1 + ((i / 2) * 4), 2,
                           1, 2);
      m_pLayout->addWidget(m_pLblsStonesLeft.last(), i + ((i / 2) * 4), 2, 1,
                           1);
      m_pLayout->addWidget(m_pLblsStoneIcon.last(), i + ((i / 2) * 4), 3, 1, 1);
      m_pLayout->addWidget(m_pLblsWon.last(), i + 1 + ((i / 2) * 4), 2, 1, 1);
      m_pLayout->addWidget(m_pLblsWinIcon.last(), i + 1 + ((i / 2) * 4), 3, 1,
                           1);

      m_pLblsPlayerName.last()->setAlignment(Qt::AlignRight);
      m_pLblsStonesLeft.last()->setAlignment(Qt::AlignRight);
      m_pLblsWon.last()->setAlignment(Qt::AlignRight);
    }
  }

  this->zoomChanged(m_pSettings->getGridSize());
  m_pFrame->setLayout(m_pLayout);
  m_pLayout->setColumnStretch(1, 1);
  m_pLayout->setColumnStretch(2, 1);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::drawPlayerIcon(const quint8 nID) {
  QPixmap iconStone(16, 16);
  iconStone.fill(m_pSettings->getBgColor());
  QPainter *pPainter = new QPainter(&iconStone);
  pPainter->setPen(QPen(Qt::black));
  pPainter->setBrush(QBrush(QColor(m_pSettings->getPlayerColor(nID))));
  pPainter->drawEllipse(0, 0, 15, 15);
  m_pLblsStoneIcon[nID]->setPixmap(iconStone);
  m_pLblsStoneIcon[nID]->setAlignment(Qt::AlignCenter);
  delete pPainter;
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::updateNames(const QStringList &sListName) {
  for (int i = 0; i < m_pSettings->getMaxNumOfPlayers(); i++) {
    if (i < sListName.size()) {
      m_pLblsPlayerName[i]->setText(sListName.at(i));
      m_pLblsPlayerName[i]->setVisible(true);
      m_pLblsStoneIcon[i]->setVisible(true);
      m_pLblsStonesLeft[i]->setVisible(true);
      m_pLblsWinIcon[i]->setVisible(true);
      m_pLblsWon[i]->setVisible(true);
    } else {
      m_pLblsPlayerName[i]->setVisible(false);
      m_pLblsStoneIcon[i]->setVisible(false);
      m_pLblsStonesLeft[i]->setVisible(false);
      m_pLblsWinIcon[i]->setVisible(false);
      m_pLblsWon[i]->setVisible(false);
    }
  }
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::zoomChanged(const int nNewGrid) {
  static bool bInitial = true;

  if (m_pSettings->getGridSize() != nNewGrid || bInitial) {
    m_pSettings->setGridSize(nNewGrid);
    this->resize(m_DefaultSize * m_pSettings->getGridSize() /
                 m_pSettings->getDefaultGridSize());
    m_pFrame->setFixedWidth(this->width());
    emit changeZoom();
    bInitial = false;
  }
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::resizeEvent(QResizeEvent *pEvent) {
  m_pFrame->setFixedWidth(pEvent->size().width());
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::startNewGame(const QString &sSavegame) {
  this->recolor();
  delete m_pGame;
  m_pGame = new Game(this);
  if (!m_pGame->createGame(sSavegame)) {
    QGraphicsScene *tmpScene = new QGraphicsScene(this);
    tmpScene->setBackgroundBrush(QBrush(m_pSettings->getBgColor()));
    m_pGraphView->setScene(tmpScene);
    this->setViewInteractive(false);
    return;
  }

  connect(m_pGame, &Game::updateNames, this, &StackAndConquer::updateNames);
  connect(m_pGame, &Game::drawIcon, this, &StackAndConquer::drawPlayerIcon);
  connect(m_pGame, &Game::updateStones, this,
          [this](const quint8 nID, const QString &sStones) {
            m_pLblsStonesLeft[nID]->setText(sStones);
          });
  connect(m_pGame, &Game::updateWon, this,
          [this](const quint8 nID, const QString &sWon) {
            m_pLblsWon[nID]->setText(sWon);
          });

  connect(m_pGame, &Game::setInteractive, this,
          &StackAndConquer::setViewInteractive);
  connect(m_pGame, &Game::highlightActivePlayer, this,
          &StackAndConquer::highlightActivePlayer);
  connect(this, &StackAndConquer::changeZoom, m_pGame, &Game::changeZoom);

  m_pGraphView->setScene(m_pGame->getScene());
  m_pGraphView->updateSceneRect(m_pGame->getScene()->sceneRect());

  QFileInfo fiBoard(m_pSettings->getBoardFile());
  this->setWindowTitle(fiBoard.baseName());

  if (!m_pGame->initCpu()) {
    this->setViewInteractive(false);
    QMessageBox::warning(this, tr("Warning"),
                         tr("An error occured during CPU initialization."));
    return;
  }

  this->setViewInteractive(true);
  m_pGame->updatePlayers(true);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::loadGame() {
  QString sFile = QFileDialog::getOpenFileName(
      this, tr("Load game"), m_userDataDir.absolutePath(),
      tr("Save games") + "(*" + Settings::SAVE_FILE_EXT + ")");
  if (!sFile.isEmpty()) {
    if (!sFile.endsWith(Settings::SAVE_FILE_EXT, Qt::CaseInsensitive)) {
      QMessageBox::warning(this, tr("Warning"), tr("Invalid save game file."));
    } else {
      this->startNewGame(sFile);
    }
  }
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::saveGame() {
  QString sFile = QFileDialog::getSaveFileName(
      this, tr("Save game"), m_userDataDir.absolutePath(),
      tr("Save games") + "(*" + Settings::SAVE_FILE_EXT + ")");
  if (!sFile.isEmpty()) {
    if (!sFile.endsWith(Settings::SAVE_FILE_EXT, Qt::CaseInsensitive)) {
      sFile += Settings::SAVE_FILE_EXT;
    }
    if (!m_pGame->saveGame(sFile)) {
      QMessageBox::warning(this, tr("Warning"), tr("Game could not be saved."));
    }
  }
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::setViewInteractive(const bool bEnabled) {
  m_pGraphView->setInteractive(bEnabled);
  m_pUi->action_SaveGame->setEnabled(bEnabled);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::highlightActivePlayer(const quint8 nActivePlayer,
                                            const quint8 nPlayerWon) {
  if (nPlayerWon > 0) {
    m_pUi->statusBar->showMessage(
        tr("%1 won the game!")
            .arg(m_pLblsPlayerName.at(nPlayerWon - 1)->text()));
    return;
  }

  for (int i = 0; i < m_pLblsPlayerName.size(); i++) {
    if ((nActivePlayer - 1) == i) {
      m_pLblsPlayerName[i]->setStyleSheet(
          QStringLiteral("color: ") +
          m_pSettings->getTextHighlightColor().name());
      m_pUi->statusBar->showMessage(
          tr("%1's turn").arg(m_pLblsPlayerName.at(i)->text()));
    } else {
      m_pLblsPlayerName[i]->setStyleSheet(QStringLiteral("color: ") +
                                          m_pSettings->getTextColor().name());
    }
  }
}

// ----------------------------------------------------------------------------
// ----------------------------------------------------------------------------

void StackAndConquer::loadLanguage(const QString &sLang) {
  if (m_sCurrLang != sLang) {
    m_sCurrLang = sLang;
    if (!StackAndConquer::switchTranslator(
            &m_translatorQt, "qt_" + sLang,
            QLibraryInfo::path(QLibraryInfo::TranslationsPath))) {
      StackAndConquer::switchTranslator(&m_translatorQt, "qt_" + sLang,
                                        m_pSettings->getSharePath() + "/lang");
    }
    if (!StackAndConquer::switchTranslator(
            &m_translator,
            ":/" + qApp->applicationName().toLower() + "_" + sLang + ".qm")) {
      StackAndConquer::switchTranslator(
          &m_translator, qApp->applicationName().toLower() + "_" + sLang,
          m_pSettings->getSharePath() + "/lang");
    }
  }
  m_pUi->retranslateUi(this);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

auto StackAndConquer::switchTranslator(QTranslator *translator,
                                       const QString &sFile,
                                       const QString &sPath) -> bool {
  qApp->removeTranslator(translator);
  if (translator->load(sFile, sPath)) {
    qApp->installTranslator(translator);
  } else {
    if (!sFile.contains(QStringLiteral("_en"))) {
      qWarning() << "Could not find translation" << sFile << "in" << sPath;
      return false;
    }
  }
  return true;
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::showRules() {
  auto *pDialog =
      new QDialog(this, this->windowFlags() & ~Qt::WindowContextHelpButtonHint);
  auto *pLayout = new QGridLayout(pDialog);
  pDialog->setWindowTitle(tr("Rules"));
  pDialog->setMinimumSize(700, 450);

  auto *pTextEditRules = new QTextEdit;
  pTextEditRules->setReadOnly(true);
  auto *pTextEditRulesATrois = new QTextEdit;
  pTextEditRulesATrois->setReadOnly(true);
  auto *pTabs = new QTabWidget(pDialog);
  pTabs->addTab(pTextEditRules, tr("Standard rules"));
  // TODO(x): Implement rules for > 2 players
  // TODO(x): Remove after implementation of > 2 players
  if (m_pSettings->getMaxNumOfPlayers() > 2) {
    pTabs->addTab(pTextEditRulesATrois, tr("Addition for > 2 players"));
  }
  auto *pCredits = new QLabel;
  pCredits->setOpenExternalLinks(true);

  pLayout->setContentsMargins(2, 2, 2, 2);
  pLayout->setSpacing(0);
  pLayout->addWidget(pTabs);
  pLayout->addWidget(pCredits);

  QString sLang(m_pSettings->getGuiLanguage());
  QFile rules(":/rules_" + sLang + ".html");
  if (!rules.exists()) {
    sLang = sLang.left(2);
    rules.setFileName(":/rules_" + sLang + ".html");
    if (!rules.exists()) {
      qWarning() << rules.fileName() << "does not exist. Loading EN fallback.";
      rules.setFileName(QStringLiteral(":/rules_en.html"));
    }
  }
  if (!rules.open(QFile::ReadOnly | QFile::Text)) {
    qWarning() << "Could not open rules:" << rules.fileName();
    QMessageBox::warning(this, tr("Warning"), tr("Could not open rules!"));
    return;
  }
  QTextStream stream(&rules);
  pTextEditRules->setHtml(stream.readAll());
  rules.close();

  sLang = m_pSettings->getGuiLanguage();
  rules.setFileName(":/rules_a_trois_" + sLang + ".html");
  if (!rules.exists()) {
    sLang = sLang.left(2);
    rules.setFileName(":/rules_a_trois_" + sLang + ".html");
    if (!rules.exists()) {
      qWarning() << rules.fileName() << "does not exist. Loading EN fallback.";
      rules.setFileName(QStringLiteral(":/rules_a_trois_en.html"));
    }
  }
  if (!rules.open(QFile::ReadOnly | QFile::Text)) {
    qWarning() << "Could not open rules a trois:" << rules.fileName();
    QMessageBox::warning(this, tr("Warning"), tr("Could not open rules!"));
    return;
  }
  stream.setDevice(&rules);
  pTextEditRulesATrois->setHtml(stream.readAll());
  rules.close();

  pCredits->setText(
      "<p>" +
      tr("These rules are licensed under Creative Commons "
         "<a href=\"https://creativecommons.org/licenses/by-nc/"
         "4.0/\">Attribution-Noncommercial 4.0 International</a> "
         "license.") +
      "<br />Designer: Dieter Stein, <a href=\"https://spielstein.com/games/"
      "mixtour/rules\">spielstein.com</a></p>");

  pDialog->show();
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::reportBug() {
  QDesktopServices::openUrl(QUrl(
      QStringLiteral("https://codeberg.org/ElTh0r0/stackandconquer/issues")));
}

// ----------------------------------------------------------------------------
// ----------------------------------------------------------------------------

void StackAndConquer::showInfoBox() {
  QMessageBox::about(
      this, tr("About"),
      QString::fromLatin1("<center>"
                          "<big><b>%1 %2</b></big><br/>"
                          "%3<br/>"
                          "<small>%4</small><br/><br/>"
                          "%5<br/>"
                          "%6<br/>"
                          "<small>%7</small>"
                          "</center><br/>"
                          "%8<br/><br/>"
                          "%9")
          .arg(qApp->applicationName(), qApp->applicationVersion(),
               QStringLiteral(APP_DESC), QStringLiteral(APP_COPY),
               "URL: <a href=\"https://codeberg.org/ElTh0r0/stackandconquer\">"
               "https://codeberg.org/ElTh0r0/stackandconquer</a>",
               tr("License") +
                   ": <a href=\"https://www.gnu.org/licenses/gpl-3.0.html\">"
                   "GNU General Public License Version 3</a>",
               tr("This application uses "
                  "<a href=\"https://invent.kde.org/frameworks/breeze-icons\">"
                  "Breeze icons from KDE</a>.") +
                   "<br/>" +
                   tr("The game is based on "
                      "<a href=\"https://spielstein.com/games/mixtour\">"
                      "Mixtour</a> by Dieter Stein."),
               "<i>" + tr("Translations") +
                   "</i><br />"
                   "&nbsp;&nbsp;- Dutch: Vistaus, Elbert Pol, maksverver<br />"
                   "&nbsp;&nbsp;- German: ElThoro<br />"
                   "&nbsp;&nbsp;- Italian: davi92<br />"
                   "&nbsp;&nbsp;- Russian: aaly11",
               "<i>" + tr("CPU players") +
                   "</i><br />"
                   "&nbsp;&nbsp;- Advanced CPU: ElThoro<br />"
                   "&nbsp;&nbsp;- Minimax & MonteCarlo CPU: maksverver"));
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::changeEvent(QEvent *pEvent) {
  if (nullptr != pEvent) {
    if (QEvent::LanguageChange == pEvent->type()) {
      m_pUi->retranslateUi(this);
      emit updateUiLang();
    }
  }
  QMainWindow::changeEvent(pEvent);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

void StackAndConquer::recolor() {
  QPixmap iconWin(QStringLiteral(":/img/trophy.png"));
  if (m_pSettings->getBgColor().lightnessF() < 0.5) {
    iconWin = QStringLiteral(":/img/trophy2.png");
  }

  for (int i = 0; i < m_pSettings->getMaxNumOfPlayers(); i++) {
    m_pLblsPlayerName[i]->setStyleSheet(QStringLiteral("color: ") +
                                        m_pSettings->getTextColor().name());
    m_pLblsStonesLeft[i]->setStyleSheet(QStringLiteral("color: ") +
                                        m_pSettings->getTextColor().name());
    m_pLblsWinIcon[i]->setPixmap(iconWin);
    m_pLblsWon[i]->setStyleSheet(QStringLiteral("color: ") +
                                 m_pSettings->getTextColor().name());
  }
}
